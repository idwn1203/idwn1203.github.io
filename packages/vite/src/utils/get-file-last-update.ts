import { spawn } from 'cross-spawn'
import dayjs from 'dayjs'

export function getFileLastUpdateTime(file: string) {
  return new Promise<string>((resolve, reject) => {
    const child = spawn('git', ['log', '-1', '--pretty="%ci"', file])
    let output = ''
    child.stdout.on('data', d => (output += String(d)))
    child.on('close', () => {
      const date = new Date(output)
      resolve(
        dayjs(date).format('YYYY/MM/DD HH:mm:ss'),
      )
    })
    child.on('error', reject)
  })
}
