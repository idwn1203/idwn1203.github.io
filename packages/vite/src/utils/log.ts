/* eslint-disable no-console */
import pc from 'picocolors'

const SVELTEPRESS_INFO_PREFIX = '[Sveltepress]'

export function info(...messages: any[]) {
  console.log(pc.bold(pc.white(pc.bgCyan(SVELTEPRESS_INFO_PREFIX))), ...messages)
}
