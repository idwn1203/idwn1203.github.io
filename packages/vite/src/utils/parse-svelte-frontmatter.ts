import { resolve } from 'node:path'
import { cwd } from 'node:process'
import { parse, preprocess } from 'svelte/compiler'
import sveltePreproces from 'svelte-preprocess'
import type { Expression, Pattern, PrivateIdentifier, SpreadElement } from 'estree'

/**
 * Parse the svelte source and get the frontmatter
 * For example
 * ```svelte
 * <script context="module">
 *   export const frontmatter = {
 *     title: 'some title',
 *     foo: 'bar'
 *   }
 * </script>
 * ```
 * Code like this would resolved as the following result
 * ```js
 * {
 *   title: 'some title',
 *   foo: 'bar'
 * }
 * ```
 * @param svelteCode the svelte source code
 */
export async function parseSvelteFrontmatter(svelteCode: string) {
  const preprocessedResult = await preprocess(svelteCode, sveltePreproces({
    tsconfigDirectory: resolve(cwd(), 'tsconfig.json'),
  }), {
    filename: 'App.svelte',
  })

  const fm: Record<string, any> = {}
  const ast = parse(preprocessedResult.code)
  if (!ast.module)
    return fm
  ast.module.content.body.forEach(line => {
    if (line.type !== 'ExportNamedDeclaration')
      return
    if (line.declaration?.type !== 'VariableDeclaration')
      return
    if (line.declaration.kind !== 'const')
      return
    const variable = line.declaration.declarations[0]
    if (!variable || variable.type !== 'VariableDeclarator')
      return
    if (variable.id.type !== 'Identifier')
      return
    if (variable.id.name !== 'frontmatter')
      return
    if (!variable.init)
      return
    if (variable.init.type !== 'ObjectExpression')
      return
    variable.init.properties.forEach(prop => {
      if (prop.type !== 'Property')
        return
      recursivelySetValue(fm, prop.key, prop.value)
    })
  })

  return fm
}

function recursivelySetValue(
  fm: Record<string, any>,
  key: Expression | PrivateIdentifier,
  value: Expression | Pattern | SpreadElement | null,
) {
  if (key.type !== 'Literal' && key.type !== 'Identifier')
    return
  const keyName = String(key.type === 'Identifier' ? key.name : key.value)
  switch (value?.type) {
    case 'Literal':
      fm[keyName] = value.value
      break
    case 'ArrayExpression':
      fm[keyName] = []
      value.elements.forEach((ele, i) => {
        recursivelySetValue(fm[keyName], { type: 'Literal', value: i }, ele)
      })
      break
    case 'ObjectExpression':
      fm[keyName] = {}
      value.properties.forEach(prop => {
        if (prop.type === 'Property')
          recursivelySetValue(fm[keyName], prop.key, prop.value)
      })
  }
}
