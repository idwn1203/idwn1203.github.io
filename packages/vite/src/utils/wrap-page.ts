import { LRUCache } from 'lru-cache'
import type { ResolvedTheme } from '../types'
import mdToSvelte from '../markdown/md-to-svelte.js'
import { parseSvelteFrontmatter } from './parse-svelte-frontmatter.js'
import { getFileLastUpdateTime } from './get-file-last-update.js'

const cache = new LRUCache<string, any>({ max: 100 })
export const scriptRe = /<script\b[^>]*>[\s\S]*?<\/script\b[^>]*>/g
const styleRe = /<style\b[^>]*>[\s\S]*?<\/style\b[^>]*>/g
const svelteHeadRe = /<svelte:head>[\s\S]*?<\/svelte:head>/g
const svelteBodyRe = /(<svelte:body\b[^>]*>[\s\S]*?<\/svelte:body>)/g
const svelteWindowRe = /(<svelte:window\b[^>]*>[\s\S]*?<\/svelte:window>)|(<window\b[^>]*\/>)/g

export async function wrapPage({
  layout,
  id,
  mdOrSvelteCode,
  highlighter,
  rehypePlugins,
  remarkPlugins,
  footnoteLabel,
}: {
  mdOrSvelteCode: string
  id: string
  layout?: string
} & Partial<Omit<ResolvedTheme, 'name' | 'vitePlugins' | 'pageLayout' | 'globalLayout'>>) {
  const cacheKey = JSON.stringify({ id, mdOrSvelteCode })
  let cached = cache.get(cacheKey)
  if (cached)
    return cached

  let fm: Record<string, any> = {}
  let svelteCode = ''

  const lastUpdate = await getFileLastUpdateTime(id)

  if (id.endsWith('.md')) {
    const { code, data } = await mdToSvelte({
      mdContent: mdOrSvelteCode,
      highlighter,
      remarkPlugins,
      rehypePlugins,
      filename: id,
      footnoteLabel,
    }) || { code: '', data: {} }
    const { fm: dataFm = {}, ...others } = data || { fm: {} }
    fm = {
      pageType: 'md',
      lastUpdate,
      ...(dataFm as any),
      ...others,
    }
    svelteCode = code
  }
  else if (id.endsWith('page.svelte')) {
    fm = {
      ...(await parseSvelteFrontmatter(mdOrSvelteCode)),
      pageType: 'svelte',
      lastUpdate,
    }
    svelteCode = mdOrSvelteCode
  }
  else if (id.endsWith('layout.svelte')) {
    svelteCode = mdOrSvelteCode
  }

  let wrappedCode = svelteCode
  if (layout) {
    wrappedCode = wrapSvelteCode({
      svelteCode,
      fm,
      pageLayout: layout,
    })
  }
  cached = {
    wrappedCode,
    fm,
  }
  cache.set(cacheKey, cached)
  return cached
}

export function wrapSvelteCode({
  pageLayout,
  svelteCode,
  fm,
}: {
  svelteCode: string
  pageLayout: string
  fm: Record<string, any>
}) {
  const imports = [
    `import PageLayout from '${pageLayout}'`,
    `const fm = ${JSON.stringify(fm)}`,
  ].join('\n')

  const svelteTagReArr = [svelteHeadRe, svelteBodyRe, svelteWindowRe]
  const svelteBuiltinTags = svelteTagReArr.reduce<string[]>((res, re) => {
    const tags = hoistTag(re, svelteCode)
    svelteCode = svelteCode.replace(re, '')
    return [
      ...res,
      ...tags,
    ]
  }, [])

  const scripts = hoistTag(scriptRe, svelteCode)
  if (scripts.length) {
    scripts[0] = scripts[0].replace(/<script\b[^>]*>/, m => [
      m,
      imports,
    ].join('\n'))
  }
  else {
    scripts.push('<script>', imports, '</script>')
  }

  const styleMatches = styleRe.exec(svelteCode)
  let styleCode = ''
  if (styleMatches) {
    styleCode = styleMatches[0]
    svelteCode = svelteCode.replace(styleRe, '')
  }

  svelteCode = svelteCode.replace(scriptRe, '')
  return `${scripts.join('\n')}
${svelteBuiltinTags.join('\n')}
<PageLayout {fm}>${svelteCode}</PageLayout>
${styleCode}
`
}

function hoistTag(tagRe: RegExp, svelteCode: string) {
  const tags = []
  let matches: RegExpMatchArray | null = null
  do {
    matches = tagRe.exec(svelteCode)
    if (matches)
      tags.push(matches[0])
  } while (matches)
  return tags
}
