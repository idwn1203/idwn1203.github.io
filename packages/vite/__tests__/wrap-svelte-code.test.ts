import { describe, expect, it } from 'vitest'
import { wrapSvelteCode } from '../src/utils/wrap-page'

const source = `
<script context="module">
  export const someVar = 'some value'
</script>
<script>
import SomeComponent from '/path/to/SomeComponent.svelte'

let innerWidth
</script>

<svelte:body on:scroll></svelte:body>

<svelte:head>
  <style>
    .foo {
      color: red;
    }
  </style>
  <script>
    var map = null
  </script>
</svelte:head>

# Title Level 1

Paragraph

<SomeComponent />

<svelte:window bind:innerWidth={innerWidth}></svelte:window>

<style>
  h1 {
    color: blue;
  }
</style>
`

describe('wrap svelte code', () => {
  it('with script, style and context script', () => {
    expect(wrapSvelteCode({
      svelteCode: source,
      fm: {
        title: 'title',
      },
      pageLayout: '/path/to/PageLayout.svelte',
    })).toMatchSnapshot()
  })
})
