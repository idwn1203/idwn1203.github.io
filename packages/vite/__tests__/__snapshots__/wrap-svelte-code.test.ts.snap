// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wrap svelte code > with script, style and context script 1`] = `
"<script context="module">
import PageLayout from '/path/to/PageLayout.svelte'
const fm = {"title":"title"}
  export const someVar = 'some value'
</script>
<script>
import SomeComponent from '/path/to/SomeComponent.svelte'

let innerWidth
</script>
<svelte:head>
  <style>
    .foo {
      color: red;
    }
  </style>
  <script>
    var map = null
  </script>
</svelte:head>
<svelte:body on:scroll></svelte:body>
<svelte:window bind:innerWidth={innerWidth}></svelte:window>
<PageLayout {fm}>







# Title Level 1

Paragraph

<SomeComponent />




</PageLayout>
<style>
  h1 {
    color: blue;
  }
</style>
"
`;
